tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -1137,6 +1137,9 @@ unsigned CGMSHLSLRuntime::ConstructStructAnnotation(DxilStructAnnotation *annota
         fieldAnnotation.SetPrecise();
     }
 
+    if (RD->isUnion())
```

Comment:
Looking at this deeper, I think this bit is a problem.

This change as it is makes unions empty structs in the DXIL StructAnnotation, which causes them to have size 0 in cbuffers, and is the first of _many_ issues that cause HLSL's SROA pass to fall over.

Once you fix that, SROA currently restricts how bitcasts can be used in ways that are fundamentally incompatible with unions. Overall I think this change is going to need a lot more baking.